### PR TITLE
Update LDAP encoders

### PR DESCRIPTION
**Pull request by noloader**
This commit updates:

* LDAP encoder documentation;
* LDAP encoding in encodeForDN and encodeForLDAP;
* LDAP tests for encodeForDN and encodeForLDAP;

### CODE REVIEW

**Comment by kwwall** [comment]
@noloader - Before I start reviewing this code in depth, please summarize why you think that we need a new `Encoder` method, `encodeForLdapSearch` and that we can't just have `encodeForLdap` do that functionality as described in RFC 4515. I don't really see that RFC as offering / suggesting that there are 3 different use cases. I think it only deals with encoding DNs / RDNs and encoding for LDAP search filters.  (The encoding for LDIF is mentioned in a different RFC, but that indeed would be a third if you wish to implement that.) The `Encoder` interface is already bloated IMO, so I'm reluctant to add new methods. Thanks. (And I vaguely recall you asking about this in an earlier email, but I didn't have time to comment on. Sorry about that.)

**Comment by noloader** [comment]
Thanks @kwwall.

>  please summarize why you think that we need a new Encoder method, encodeForLdapSearch and that we can't just have encodeForLdap do that functionality as described in RFC 4515. 

I removed that extra encoder.

I also rebased and squashed against the latest tip of develop.

**Comment by noloader** [comment]
@kwwall,

I've been thinking about the old CSRSS backspace bug. In the old days, it was possible to crash Windows `csrss.exe` with a string like `test\b\b\b\b\b\b\b\`. It would cause an out-of-bounds memory access.

Do you think it would be a good idea to encode whitesapce characters and backspace (except traditional SPACE)?

RFC 4514, Section 4, shows an example of escaping the `CR`:

> CN=Before\0dAfter,DC=example,DC=net

**Comment by kwwall** [comment]
Suggest maybe adding a new property something like:
    # Comma-separated list of hex values that should also be encoded in
additon to those mentioned in RFC 4515:
     # Encode backspace and CR.
Encoder.LDAPSearch.extraCharsToEncode=0x08,0x0d

And then if such a property value is non-empty, you encode those
characters. I figured using hex values is better than using \b or \r
because not all hex values have such shortcuts.

Also, this approach is more general and puts the control back in the hands
of the application developers, where it belongs.  That seems a lot better
to me than hard-coding a few select characters.
.
P.S. - No need to rebase and squash. When I do the merge, I usually pick
"Squash and merge", which has a similar effect.

File: src/main/java/org/owasp/esapi/Encoder.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/Encoder.java:452]: Suggest changing 'NB:' to 'NOTE:' (or perhaps 'WARNING:' if you think it's really important to mention because it is likely to break someone's code.

Outside of legal notation, NB is not commonly used any longer, at least not like it was 30 years ago or so when I used to see it in comments throughout UNIX code bases. I'm not sure younger developers will know what that NB stands for the Latin _nota bene_, commonly translated as "note well" unless perhaps they took Latin in high school.

File: src/main/java/org/owasp/esapi/Encoder.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/Encoder.java:475]: Same here. Note that this is only a suggestion, but there's no point in making this obtuse and the most common meanings today for the abbreviation "NB" is no longer _nota bene_.

File: src/main/java/org/owasp/esapi/reference/DefaultEncoder.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/reference/DefaultEncoder.java:354]: Suggest changing this to:
```java
    ;    // Ignore; UTF-8 is always supported
```
That way, it makes it a little more obvious of the intent when someone sees a\this (false positive) reported by a SAST tool. (Unfortunately, many of them still report this.)

File: src/main/java/org/owasp/esapi/reference/DefaultEncoder.java
Comment (kwwall) [inline src/main/java/org/owasp/esapi/reference/DefaultEncoder.java:415]: Ditto.

File: src/test/java/org/owasp/esapi/reference/EncoderTest.java
Comment (kwwall) [inline src/test/java/org/owasp/esapi/reference/EncoderTest.java:1]: If you still intend to add encoding for CR and LF, we will need additional tests for that.

**Review by kwwall — APPROVED** [review]
This looks fine, but I left a few minor suggestions.
I'm not going to merge yet though, because I'm not sure if you wanted to do the encoding for CR and LF like you suggested (or, generalizing that, like I recommended -- so it could be disabled) or want to incorporate the suggestions to some of the comments I made here as part of this review. Therefore I will wait to merge. Once you are satisfied with all your changes, just @-me with comment you are ready to merge.

**Comment by kwwall** [comment]
@xeno6696 - If you want to review this PR, that's fine but not strictly necessary. But if you do, please hold off on the merging (assuming that you approve in the code review) until we hear back from @noloader to see if he intends to make further changes. I'd rather make related changes as part of this PR than having to create a new PR.

Also, @noloader - could you please create a GitHub Issue for this and associate it to this PR. It can be something really simple like "Ensure strict conformance with section 3 of RFC 4515 in terms of character ranges" or more detailed, depending on your tastes. That will help me when I prepare the release notes.

**Comment by kwwall** [comment]
@noloader - did you intend to add to this PR to discuss the reply I made in https://github.com/ESAPI/esapi-java-legacy/pull/790#issuecomment-1596484323 or are you through?

@xeno6696 - We need one more person to approve this. Can you look at it?

**Comment by kwwall** [comment]
@xeno6696 or @jeremiahjstacey - We need one more person to approve this. Can one of you PLEASE look at this ASAP. I'd like to get it into the upcoming ESAPI 2.5.3.0 release.

File: src/main/java/org/owasp/esapi/reference/DefaultEncoder.java
Comment (xeno6696) [inline src/main/java/org/owasp/esapi/reference/DefaultEncoder.java:349]: Something feels a bit off here.  char is 16bit.... but the data in `input` could be a string of characters that won't live through this translation.  I'm thinking of my unit tests over here:

```
    @Test
    public void test32BitCJK(){
        String s = "𡘾𦴩𥻂";
        String expected = "&#x2163e;&#x26d29;&#x25ec2;";
        String bad = "&#xd845;&#xde3e;&#xd85b;&#xdd29;&#xd857;&#xdec2;";
        assertEquals(false, expected.equals(bad));
        assertEquals(expected, codec.encode(new char[0], s));
    }

    @Test
    public void test32BitCJKMixedWithBmp(){
        String s = "𡘾𦴩<𥻂";
        String expected = "&#x2163e;&#x26d29;&lt;&#x25ec2;";
        String bad = "&#xd845;&#xde3e;&#xd85b;&#xdd29;&#xd857;&#xdec2;";
        assertEquals(false, expected.equals(bad));
        assertEquals(expected, codec.encode(new char[0], s));
    }
```
    
So imagine that the input is `𡘾𦴩𥻂`

I would think that there would be an overflow the second we call `char c = input.charAt(i);`

Because under the hood, java is representing `𡘾𦴩𥻂` as a UTF-16 buffer.  `𡘾` would get translated to garbage and then disaster follows.

File: src/main/java/org/owasp/esapi/reference/DefaultEncoder.java
Comment (xeno6696) [inline src/main/java/org/owasp/esapi/reference/DefaultEncoder.java:410]: Same comment here.

**Review by xeno6696 — COMMENTED** [review]
@kwwall @noloader If the LDAP standard has been expanded to accomodate UTF-8 encoded bytes then our API for LDAP encoding needs to be broken out into Codecs so that way we can at least maintain legacy functionality by leaving the old behavior alone and then more carefully considering the actual use case of UTF-8 bytes.  

I don't feel comfortable approving this because as written, this directly and permanently affects the encodeForLDAP method.

File: src/main/java/org/owasp/esapi/reference/DefaultEncoder.java
Comment (kwwall) [reply src/main/java/org/owasp/esapi/reference/DefaultEncoder.java:349]: @xeno6696 - Can this wait until _after_ this upcoming 2.5.3.0 release? In other words, I'll phrase the question this way: Which is less wrong? The way the code presently is or the way this PR is? If you honestly think that the code in this PR is worse than what we presently have, we can exclude it from this next release, but I'm not sure I see it this way. Even if this code has the issues that you mention, I think it moves us toward the direction of the of compliance with the RFC. So, let's not let the perfect become the enemy of the good here. IMO, unless this is going to open up worse security holes in your opinion, I think the way forward should create a GitHub issue as to what needs fixed once this PR is merged and then merge this and deal with the rest of it later.

@noloader, @xeno6696 - Thoughts?

**Comment by xeno6696** [comment]
> @noloader - Before I start reviewing this code in depth, please summarize why you think that we need a new `Encoder` method, `encodeForLdapSearch` and that we can't just have `encodeForLdap` do that functionality as described in RFC 4515. I don't really see that RFC as offering / suggesting that there are 3 different use cases. I think it only deals with encoding DNs / RDNs and encoding for LDAP search filters. (The encoding for LDIF is mentioned in a different RFC, but that indeed would be a third if you wish to implement that.) The `Encoder` interface is already bloated IMO, so I'm reluctant to add new methods. Thanks. (And I vaguely recall you asking about this in an earlier email, but I didn't have time to comment on. Sorry about that.)

I didn't realize that @noloader was implementing your suggestion here.  It's kinda late for me to disagree with it but my strategy would have been to create a new method with the better logic and then followed the deprecation strategy.  I don't see the need to force a block here @kwwall and I think I need to develop a weekly routine to check up on things so I don't hold them up.

File: src/main/java/org/owasp/esapi/reference/DefaultEncoder.java
Comment (xeno6696) [reply src/main/java/org/owasp/esapi/reference/DefaultEncoder.java:349]: My main worry is this:  Assuming that the rules we were breaking were breaking because (like browsers) some LDAP systems don't obey the RFC to the letter anyway, we could unintentionally break something.  That said, I'm also unsure of how many folks are using ESAPI for this anyway.  I won't stand in the way of the merge and I'm okay with punting for future issues, but then you also stated a desire to push 2.x into maintenance only which effectively bars that.  

Just putting that out there.

**Pull request merged**
